app/display.c

```diff
@@ -49,7 +49,7 @@
 
 static const char spin_state[NUM_SPIN_STATES] = { '|', '/', '-', '\\' };
 
-static const char *cpu_mode_str[] = { "PAR", "SEQ", "RR " };
+static const char cpu_mode_str[3][4] = { "PAR", "SEQ", "RR " };
```

Comment:
Just for my knowledge, what's the point of doing this? Is it just to avoid the array of pointers and reduce the size of the code by 12 bytes in 64-bit mode?

Reply:
Correct, the point is to avoid the array of pointers (for storing strings which were already padded to the same length) and therefore especially the relocations for these pointers, which take up significantly more space than the pointers and strings:

```
git log --oneline | head -1; cd build32; make clean &> /dev/null; make -j &> /dev/null; size app/config.o memtest_shared; objdump -x memtest_shared | grep '\.rela\.'; cd ..; cd build64; make clean &> /dev/null; make -j &> /dev/null; size app/config.o memtest_shared; objdump -x memtest_shared | grep '\.rela\.'; cd ..
6d11069 Move more of the simple string functions to the header, to allow inlining and further optimization.
   text    data     bss     dec     hex filename
  10105      19    2712   12836    3224 app/config.o
 106580   26720   13344  146644   23cd4 memtest_shared
   text    data     bss     dec     hex filename
   8653      19    2712   11384    2c78 app/config.o
 110969   28392  294688  434049   69f81 memtest_shared
  7 .rela.data    00001368  0000000000019e78  0000000000019e78  0001ae78  2**3
  8 .rela.got     000000c0  000000000001b1e0  000000000001b1e0  0001c1e0  2**3

git log --oneline | head -1; cd build32; make clean &> /dev/null; make -j &> /dev/null; size app/config.o memtest_shared; objdump -x memtest_shared | grep '\.rela\.'; cd ..; cd build64; make clean &> /dev/null; make -j &> /dev/null; size app/config.o memtest_shared; objdump -x memtest_shared | grep '\.rela\.'; cd ..
88843af Add support for configuring the CPU sequencing mode through the command line. Reorder tests in app/config.c::parse_option alphabetically. Fixes #82.
   text    data     bss     dec     hex filename
  10266      19    2712   12997    32c5 app/config.o
 106705   26720   13344  146769   23d51 memtest_shared
   text    data     bss     dec     hex filename
   8811      19    2712   11542    2d16 app/config.o
 111059   28360  294720  434139   69fdb memtest_shared
  7 .rela.data    00001320  0000000000019f18  0000000000019f18  0001af18  2**3
  8 .rela.got     000000c0  000000000001b238  000000000001b238  0001c238  2**3
```

On the debrouxl/experiments branch, there's a commit containing this change and two other one-liners switching from array of pointers to strings which were already padded to the same length to an array of fixed-length strings. The combined effect is stronger:
```
git log --oneline | head -1; cd build32; make clean &> /dev/null; make -j &> /dev/null; size app/config.o memtest_shared; objdump -x memtest_shared | grep '\.rela\.'; cd ..; cd build64; make clean &> /dev/null; make -j &> /dev/null; size app/config.o memtest_shared; objdump -x memtest_shared | grep '\.rela\.'; cd ..
d1c332b Further optimize the JEP106 list by switching from relocated string pointers to offsets from the beginning of a string table.
   text    data     bss     dec     hex filename
  10374      19    2712   13105    3331 app/config.o
 106377   25636   13408  145421   2380d memtest_shared
   text    data     bss     dec     hex filename
   8734      19    2712   11465    2cc9 app/config.o
 108873   26256  294784  429913   68f59 memtest_shared
  7 .rela.data    000006d8  000000000001a2c0  000000000001a2c0  0001b2c0  2**3
  8 .rela.got     000000d8  000000000001a998  000000000001a998  0001b998  2**3

git log --oneline | head -1; cd build32; make clean &> /dev/null; make -j &> /dev/null; size app/config.o memtest_shared; objdump -x memtest_shared | grep '\.rela\.'; cd ..; cd build64; make clean &> /dev/null; make -j &> /dev/null; size app/config.o memtest_shared; objdump -x memtest_shared | grep '\.rela\.'; cd ..
1b339de Optimize two arrays of strings and an array of structs containing same-, fixed-length strings by switching to fixed-length char arrays, which avoids pointers + relocations.
   text    data     bss     dec     hex filename
  10374      19    2712   13105    3331 app/config.o
 105792   25988   13312  145092   236c4 memtest_shared
   text    data     bss     dec     hex filename
   8731      19    2712   11462    2cc6 app/config.o
 107992   26512  294784  429288   68ce8 memtest_shared
  7 .rela.data    00000528  000000000001a100  000000000001a100  0001b100  2**3
  8 .rela.got     000000d8  000000000001a628  000000000001a628  0001b628  2**3
```

I did disassemble the binary to check the generated code and data :)